exonum/src/runtime/error/execution_error.rs

```diff
@@ -12,30 +12,238 @@
+use exonum_proto::ProtobufConvert;
+use failure::bail;
 
-pub fn serialize<S>(inner: &ExecutionError, serializer: S) -> Result<S::Ok, S::Error>
-where
-    S: Serializer,
-{
-    ExecutionStatus::from(Err(inner)).serialize(serializer)
+use std::{
+    any::Any,
+    convert::TryFrom,
+    fmt::{self, Display},
+};
+
+use super::{execution_status::serde::ExecutionStatus, ErrorKind, ErrorMatch, ExecutionError};
+use crate::{
+    crypto::{self, Hash},
+    proto::schema::runtime as runtime_proto,
+    runtime::{CallSite, RuntimeIdentifier},
+};
+
+/// Custom `serde` implementation for `ExecutionError`.
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct ExecutionErrorSerde;
+
+impl ExecutionError {
+    /// Creates a new execution error instance with the specified error kind
+    /// and an optional description.
+    pub fn new(kind: ErrorKind, description: impl Into<String>) -> Self {
+        Self {
+            kind,
+            description: description.into(),
+            runtime_id: None,
+            call_site: None,
+        }
+    }
+
+    /// Creates an execution error for use in service code.
+    pub fn service(code: u8, description: impl Into<String>) -> Self {
+        Self::new(ErrorKind::Service { code }, description)
+    }
+
+    /// Creates an execution error from the panic description.
+    pub(super) fn from_panic(any: impl AsRef<(dyn Any + Send)>) -> Self {
+        let any = any.as_ref();
+
+        // Tries to get a meaningful description from the given panic.
+        let description = if let Some(s) = any.downcast_ref::<&str>() {
+            s.to_string()
+        } else if let Some(s) = any.downcast_ref::<String>() {
+            s.clone()
+        } else if let Some(error) = any.downcast_ref::<Box<(dyn std::error::Error + Send)>>() {
+            error.description().to_string()
+        } else if let Some(error) = any.downcast_ref::<failure::Error>() {
+            error.to_string()
+        } else {
+            // Unknown error kind; keep its description empty.
+            String::new()
+        };
+
+        Self::new(ErrorKind::Unexpected, description)
+    }
+
+    /// Converts an error to a matcher. The matcher expect the exact kind and description
+    /// of this error, and does not check any other error fields.
+    pub fn to_match(&self) -> ErrorMatch {
+        ErrorMatch::new(self.kind, self.description.clone())
+    }
+
+    /// The kind of error that indicates in which module and with which code the error occurred.
+    pub fn kind(&self) -> ErrorKind {
+        self.kind
+    }
+
+    /// Human-readable error description. May be empty.
+    pub fn description(&self) -> &str {
+        &self.description
+    }
+
+    /// Returns the ID of a runtime in which this error has occurred. If the runtime is not known
+    /// (e.g. error originates in the core code), returns `None`.
+    pub fn runtime_id(&self) -> Option<u32> {
+        self.runtime_id
+    }
+
+    #[inline]
```

Comment:
What is reason to use `inline` directive explicitly?

Reply:
Dunno, it was there before, I left it as-is.